docs/proposal/20240308-vpc-infra-creation.md

```diff
@@ -0,0 +1,393 @@
+# Dynamically create infrastructure required for VPC cluster
+
+## Motivation
+Currently, inorder to create a VPC cluster using cluster api we need to create
+resources as prerequisites. This involves work which is limiting the true
+capabilities of cluster api.
+
+
+## Goal
+1. Dynamically create the required cloud resources as a part of cluster
+   creation.
+
+
+## Proposal
+
+
+### Cluster API VPC cluster components
+
+![vpc-cluster-components.png](../images/vpc-cluster-components.png)
+
+
+### Proposed API changes
+
+
+#### VPCCluster improvements
+```go
+// IBMVPCClusterSpec defines the desired state of IBMVPCCluster.
+type IBMVPCClusterSpec struct {
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// controlPlaneEndpoint represents the endpoint used to communicate with the control plane.
+	// +optional
+	ControlPlaneEndpoint capiv1beta1.APIEndpoint `json:"controlPlaneEndpoint"`
+
+	// controlPlaneLoadBalancer is optional configuration for customizing control plane behavior.
+	// +optional
+	ControlPlaneLoadBalancer []*VPCLoadBalancerSpec `json:"controlPlaneLoadBalancer,omitempty"`
+
+	// cosInstance is the IBM COS instance to use for cluster resources.
+	COSInstance *CosInstance `json:"cosInstance,omitempty"`
+
+	// networkSpec represents the VPC network to use for the cluster.
+	NetworkSpec *VPCNetworkSpec `json:"networkSpec,omitempty"`
+
+	// region defines the IBM Cloud Region the cluster resources will be deployed in.
+	Region string `json:"region"`
+
+	// resourceGroup defines the IBM Cloud resource group where the cluster resources should be created.
+	ResourceGroup string `json:"resourceGroup"`
+
+	// The Name of VPC. 
+	// Deprecated: use NetworkSpec instead.
+	VPC string `json:"vpc,omitempty"`
+
+	// The Name of availability zone.
+	// Deprecated: use NetworkSpec instead.
```

Comment:
I don't see any field indicating `zone` data in `NetworkSpec`, how is that being handled once we deprecate this.
cc @Karthik-K-N

Reply:
Zones will be defined within sub-resources. VPC isn't restricted to a single zone, it can deploy (or use) Subnets, VSI's, etc. across multiple zones, of which I expect those resources to specify their Zone.